src/filter/http_context.rs

```diff
@@ -1,159 +1,259 @@
-use crate::configuration::FilterConfig;
+use crate::configuration::{FilterConfig, RateLimitPolicy, Rule};
 use crate::envoy::{
-    RLA_action_specifier, RateLimitRequest, RateLimitResponse, RateLimitResponse_Code,
+    RLA_action_specifier, RateLimitDescriptor, RateLimitDescriptor_Entry, RateLimitRequest,
+    RateLimitResponse, RateLimitResponse_Code,
 };
-use crate::utils::{descriptor_from_actions, request_process_failure, UtilsErr};
+use crate::utils::{match_headers, request_process_failure, subdomain_match};
 use log::{debug, info, warn};
 use protobuf::Message;
 use proxy_wasm::traits::{Context, HttpContext};
 use proxy_wasm::types::Action;
+use std::collections::HashMap;
 use std::rc::Rc;
 use std::time::Duration;
 
 const RATELIMIT_SERVICE_NAME: &str = "envoy.service.ratelimit.v3.RateLimitService";
 const RATELIMIT_METHOD_NAME: &str = "ShouldRateLimit";
 
-// we cannot determine using Istio's naming scheme because this cluster is patched
-// by the kuadrant-controller and should match whatever value is patched.
-const DEFAULT_UPSTREAM_CLUSTER: &str = "rate-limit-cluster";
-
-struct RequestInfo {
-    pub host: String,
-    pub path: String,
-    pub method: String,
-}
-
 pub struct Filter {
     pub context_id: u32,
     pub config: Rc<FilterConfig>,
 }
 
 impl Filter {
-    fn config(&self) -> &FilterConfig {
-        &self.config
+    fn request_path(&self) -> String {
+        // Note(eastizle): self.get_http_request_header(":path") is equivalent?
```

Comment:
Worthwhile having a look at, the former would be nice… avoids some work from the `vec!`